Add Tilt::Template#freeze_string_literals?

jeremyevans opened the pull request:
If this returns true, frozen-string-literal: true is set when
compiling the template method.  It is false by default, and
will need to be manually set to true in the template classes
that could support it.

jeremyevans commented:
I've rebased this against the master branch, and added support for the Erubi template to use this.  This can decrease strings allocated during template rendering, and allows for additional optimizations in Erubi (avoiding the explicit use of `.freeze` for literal strings, which improves performance, see https://github.com/jeremyevans/erubi/pull/33). @judofyr this has been open for over 5 years now, could you please review when you have time?

jeremyevans commented:
@judofyr This is something I plan to merge unless you object.

judofyr commented:
If this is *only* going to be enabled by the Erubi implementation, can this instead be solved by overriding `precompiled_preamble` and adding it there? I'm just not such a big fan of introducing a new public API when we only have one known use case.

jeremyevans commented:
I added support for Erubi because that is what I use and probably the most common template library since most Ruby web applications default to erb for templates, and Tilt defaults to Erubi for erb.  However, I'm guessing this could be supported by many/most of tilt template libraries that support recompilation.  If you prefer, I can update this pull request with support for additional template libraries before merging.

jeremyevans commented:
I've rebased this against master, and added support for all 3 erb engines.  I can certainly work on other engines later, but I think this shows it is possible to support this feature in multiple template engines.

jeremyevans commented:
I also rebased this after the switch to minitest/spec.

judofyr commented:
LGTM! This is now conflicting again with `master`. If you rebase I'll merge it.

jeremyevans commented:
@judofyr I've rebased this against master.

jeremyevans commented:
@judofyr Is it possible to merge this and other pull requests (#380, #385, #387), then release 2.1.0?  I also think we should consider moving the repository to a separate GitHub organization or account where we can have control over the repository.
